ouroboros-network/src/Ouroboros/Network/Protocol/ChainSync/ClientPipelined.hs

```diff
@@ -11,17 +11,24 @@ module Ouroboros.Network.Protocol.ChainSync.ClientPipelined
   , ClientStNext (..)
   , ClientPipelinedStIntersect (..)
   , ChainSyncInstruction (..)
+  , TraceChainsyncClientObservation (..)
 
   , chainSyncClientPeerPipelined
   , chainSyncClientPeerSender
   ) where
 
+import Control.Tracer (Tracer, traceWith)
 import Network.TypedProtocol.Core
 import Network.TypedProtocol.Pipelined
 
 import Ouroboros.Network.Protocol.ChainSync.Type
+import Ouroboros.Network.Util.TaggedObserve
 
 
+data TraceChainsyncClientObservation = IntersectObservation ObserveIndicator
+                                     | NextObservation ObserveIndicator
```

Comment:
Ah, this is an observation for `RequestNext`, not just the "next observation".

That wasn't obvious to me. Would you add comments for each of the constructors, explain what events they correspond to? (I realize this is a Draft PR, and you might already be planning to do that 👍.)

Reply:
I will document the PR before the goes out of draft. However it is still very much a work in progress, together with the `karknu/counter` branch in cardano-node it only works correctly with a single upstream client.